documentation/rir.md

```diff
@@ -0,0 +1,97 @@
+# RIR
+
+## Overview
+
+RIR is a bytecode, similar to the JVM bytecode. It stands for "R Intermediate Representation".
+
+R is an "interpreted" language, in that every statement of code is executed immediately when it's parsed. However, re-parsing every statement is slow. So, when the RIR interpreter parses code, it first compiles it into "RIR bytecode", *then* immediately executes the bytecode. When a call to previously-parsed code is encountered, like in a loop or function call, instead of re-parsing the code, RIR will simply run the existing bytecode. This is called JIT ("Just in Time") compilation.
+
+### What RIR bytecode looks like
+
+To roughly see a function's bytecode, run `PIR_DEBUG=PrintPirAfterOpt ./bin/R`, then type `f <- rir.compile(<function>)`, then `pir.compile(f)`. Here's an example:
+
+<pre>
+&gt; f &lt;- rir.compile(function() {
```

Comment:
In Markdown, instead of use `<pre>` and &gt;` you can use three backticks, e.g.:

<pre>
```
f <- rir.compile(function() ...)
```
</pre>

Or you could indent everything with 4 spaces:
<pre>
    f <- rir.compile(function() ...)
</pre>